instrumentor/controllers/instrumentationdevice/common.go

```diff
@@ -157,46 +148,46 @@ func removeInstrumentationDeviceFromWorkload(ctx context.Context, kubeClient cli
 		return errors.New("unknown kind")
 	}
 
-	err := kubeClient.Get(ctx, client.ObjectKey{
-		Namespace: namespace,
-		Name:      workloadName,
-	}, workloadObj)
-	if err != nil {
-		return client.IgnoreNotFound(err)
-	}
-
-	result, err := controllerutil.CreateOrPatch(ctx, kubeClient, workloadObj, func() error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		err := kubeClient.Get(ctx, client.ObjectKey{
+			Namespace: namespace,
+			Name:      workloadName,
+		}, workloadObj)
+		if err != nil {
+			return client.IgnoreNotFound(err)
+		}
 
-		// clear old ebpf instrumentation annotation, just in case it still exists
-		clearInstrumentationEbpf(workloadObj)
 		podSpec, err := getPodSpecFromObject(workloadObj)
 		if err != nil {
 			return err
 		}
 		// If instrumentation device is removed successfully, remove odigos.io/inject-instrumentation label to disable the webhook
-		instrumentation.RemoveInjectInstrumentationLabel(podSpec)
-
-		instrumentation.RevertInstrumentationDevices(podSpec)
-
-		err = instrumentation.RevertEnvOverwrites(workloadObj, podSpec)
+		webhookLabelRemoved := instrumentation.RemoveInjectInstrumentationLabel(podSpec)
+		deviceRemoved := instrumentation.RevertInstrumentationDevices(podSpec)
+		envChanged, err := instrumentation.RevertEnvOverwrites(workloadObj, podSpec)
 		if err != nil {
 			return err
 		}
 
-		return nil
-	})
+		// if we didn't change anything, we don't need to update the object
+		// skip the api-server call, return no-op and skip the log message
+		if !webhookLabelRemoved && !deviceRemoved && !envChanged {
+			return nil
```

Comment:
Maybe we should return here a custom error indicating that nothing has been done?
This is for logging below - to avoid logging for un-relevant workloads.

Another option is to set a bool `updatedWorkload` in the `removeInstrumentationDeviceFromWorkload` scope - initialize it to true - here we will set it to false. This will allow us to know if nil is returned because of a successful update or because of this case.